shard-operator/src/main/java/com/redhat/service/bridge/shard/operator/BridgeIngressServiceImpl.java

```diff
@@ -80,26 +84,31 @@ public Deployment fetchOrCreateBridgeIngressDeployment(BridgeIngress bridgeIngre
         environmentVariables.add(new EnvVarBuilder().withName(KafkaConfigurationCostants.KAFKA_CLIENT_SECRET_ENV_VAR).withValue(kafkaConfigurationProvider.getSecret()).build());
         environmentVariables.add(new EnvVarBuilder().withName(KafkaConfigurationCostants.KAFKA_SECURITY_PROTOCOL_ENV_VAR).withValue(kafkaConfigurationProvider.getSecurityProtocol()).build());
         environmentVariables.add(new EnvVarBuilder().withName(Constants.BRIDGE_INGRESS_BRIDGE_ID_CONFIG_ENV_VAR).withValue(bridgeIngress.getSpec().getId()).build());
-        deployment.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(environmentVariables);
+        expected.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(environmentVariables);
+
+        Deployment existing = kubernetesClient.apps().deployments().inNamespace(bridgeIngress.getMetadata().getNamespace()).withName(bridgeIngress.getMetadata().getName()).get();
 
-        return kubernetesClient.apps().deployments().inNamespace(bridgeIngress.getMetadata().getNamespace()).create(deployment);
+        if (existing == null || !DeploymentSpecUtils.isDeploymentEqual(expected, existing)) {
+            return kubernetesClient.apps().deployments().inNamespace(bridgeIngress.getMetadata().getNamespace()).createOrReplace(expected);
+        }
+
+        return existing;
     }
 
-    // TODO: if the retrieved resource spec is not equal to the expected one, we should redeploy https://issues.redhat.com/browse/MGDOBR-140
     @Override
     public Service fetchOrCreateBridgeIngressService(BridgeIngress bridgeIngress, Deployment deployment) {
-        Service service = kubernetesClient.services().inNamespace(bridgeIngress.getMetadata().getNamespace()).withName(bridgeIngress.getMetadata().getName()).get();
+        Service expected = templateProvider.loadBridgeIngressServiceTemplate(bridgeIngress);
 
-        if (service != null) {
-            return service;
-        }
+        // Specs
+        expected.getSpec().setSelector(new LabelsBuilder().withAppInstance(deployment.getMetadata().getName()).build());
 
-        service = templateProvider.loadBridgeIngressServiceTemplate(bridgeIngress);
+        Service existing = kubernetesClient.services().inNamespace(bridgeIngress.getMetadata().getNamespace()).withName(bridgeIngress.getMetadata().getName()).get();
 
-        // Specs
-        service.getSpec().setSelector(new LabelsBuilder().withAppInstance(deployment.getMetadata().getName()).build());
+        if (existing == null || !expected.getSpec().equals(existing.getSpec())) {
```

Comment:
Minor:
Service spec contains clusterIP, which is added by Kubernetes/OpenShift. Therefore equals will always be false here. Though it doesn't seem to cause any issue apart from unnecessary client call.

Reply:
The most incredible thing about Java `equals` is that we can control what to compare. In this case, we are only comparing the attributes that Service cares about. I haven't seen the code yet, but my guess would be that.

Reply:
Yeah, when I implemented this I checked the fabric8 override for `equals` and they are checking (recursively) that the object is of course exactly the same. The problems are the properties set by kubernetes... I had to implement a quite specific `equals` check for the `Deployment` few line below for this reason. 
I will move to a more specific check also for the `Service`, thanks for the catch :)